### ISSUE TITLE
trailing profit optimisation

### ISSUE THREAD

**Issue by aristosv** [freqtrade/freqtrade#4564]
## Describe your environment
Operating system: Debian 10
Python Version: Python 3.7.3
CCXT version: ccxt==1.42.66
Freqtrade Version: freqtrade develop-7e45af97
  
## Your question
My minimal_roi is set to 0.01 (1%) but I've had cases where the bot sold at 4% or 5% profit. But that happens very rarely. Obviously there's a "trailing profit" feature, and I was wondering if I could somehow optimise it, so the bot gives a better chance for higher percentage, instead of immediately selling once the minimal_roi is reached.

I want to keep it at 1%, but allow more flexibility, if the % has a chance of going higher.

**Comment by xmatthias** [comment]
you can increase / decrease roi over time, allowing you to maybe require 5% profit for the first hour, then drop to 3% , ... 

You can use the trailing stoploss feature as well - but you'll always be giving up the "top %" .
you can't set trailing-stop too tight, otherwise it'll have backtesting inflated (without ever being able to replicate this in forward testing), so you'd probably use something like an offset of 0.02, and a trailing stop fo 0.01 ... which would sell at 1% or higher - BUT requires price to first reach 2% (so you can also miss some sells).

**Comment by aristosv** [comment]
> you can increase / decrease roi over time, allowing you to maybe require 5% profit for the first hour, then drop to 3%

The problem with this, is that if I hardcode a requirement of 5% (or even 2%) in the first hour, I might miss a 4% profit, and then wait for recovery again.



> You can use the trailing stoploss feature as well

Now this, I didn't know about. It's a bit confusing calling it "stoploss" because my brain automatically assumes it's a feature that stops the loss. But then you add "[with positive offset](https://www.freqtrade.io/en/latest/strategy-advanced/#trailing-stoploss-with-positive-offset)" and it makes sense again. :) (my 2c, maybe it would be better if it was called trailing profit?)

I will look into this, thanks for the info.

**Comment by jej** [comment]
> maybe it would be better if it was called trailing profit

depending of the point of view I guess... it is still triggered regarding a loss.

**Comment by xmatthias** [comment]
also on the exchanges (which support it - like FTX), this type of order is called a trailing stoploss order - which automatically trails the price with X% offset.

**Comment by aristosv** [comment]
I am trying to get a better understanding of how this works, so please bare with me. If it's preferable to raise a new question, let me know.
My target is 1% profit, and then trail this profit if it goes higher, for as far as it will go. If it makes a .25% drop (after it went above 1%), sell it.

```
        if current_profit < 0.01:
            return -1 # so this tells me that as long as the trade is below 1% profit, wait.

        # After reaching 1% profit, follow it, in 0.25% increments
        desired_stoploss = current_profit / 4

        # Use a minimum of 2.5% and a maximum of 5%
        return max(min(desired_stoploss, 0.05), 0.025) # This I don't really understand what it does. Does it set a maximum of 5% profit, and a minimum of 2.5% above the already gained profit of 1%?
```

**Comment by xmatthias** [comment]
0.25% drop is too tight for sure.

what you're doing however is something different - `desired_stoploss = current_profit / 4` will increase the gap with raising profit



so for 1% profit, it'll be 0.25%
for 10% profit it'll be 2.5% 
...


The function you provide above increases the gap between price and stoploss as profits raise (giving it more room to eventually pull back), and the last line simply limits this gap to be between 2.5% and 5% (overwriting your 0.25% initial stoploss with 2.5%).
The explanation for this is [here](https://www.freqtrade.io/en/latest/strategy-advanced/#trailing-stoploss-with-positive-offset) - and it includes a full description.


In my opinion, to simply trail with 0.25% after 1% profit is as simple as the following.
```
    trailing_stop = True
    trailing_stop_positive = 0.0025    # This is too tight!
    trailing_stop_positive_offset = 0.01
    trailing_only_offset_is_reached = True
```

**Comment by pampos79** [comment]
Is it better to have our trailing_stop_positive at 0.5%? Lets say i have an open trade for 50 USD and at 1% profit what is the best trailing_stop_positive percentage to use?

**Comment by xmatthias** [comment]
in general, i'd not use a trailing stop below the average candle movements (or below 1%). 

using values below the average candle range will heavily inflate backtesting profits - and anything below 1% is usually problematic when placing stoplosses on the exchange.

**Comment by aristosv** [comment]
I think the best way to demonstrate what I want to do is by absolute stoploss

```
if current_profit > 0.2:
    return (-0.1 + current_profit)
if current_profit > 0.3:
    return (-0.2 + current_profit)
if current_profit > 0.4:
    return (-0.3 + current_profit)
if current_profit > 0.5:
    return (-0.4 + current_profit)
if current_profit > 0.6:
    return (-0.5 + current_profit)
if current_profit > 0.7:
    return (-0.6 + current_profit)
if current_profit > 0.8:
    return (-0.7 + current_profit)
if current_profit > 0.9:
    return (-0.8 + current_profit)
and so on and so on....
return 1
```

I just want to have a safety net, just below the current profit. I think a 1% or even a 0.5% safety net would be fine.

**Comment by aristosv** [comment]
Maybe absolute stoploss is not the best way to implement this though.

**Comment by xmatthias** [comment]
no, it's absolutely not - as posted above - you do NOT need custom stoploss to trail current price with 1%.
the configuration i posted above will do exactly that (after tuning the values, obviously).

**Comment by pampos79** [comment]
I want to set a minimal_roi of 1.1% and then monitor the profit by 0.1% increments.

For example, when a trade reaches a 1.1% profit, I want the bot to start monitoring. As long as it’s going up (1.2%, 1.3%, 1.4% etc.), I want the bot to let it keep going up for as long as it will go.

As soon as the profit drops by 0.1% (e.g., from 1.4% to 1.3%) immediately sell.

Is this possible?

**Comment by xmatthias** [comment]
0.1% is too tight and will not work.
i understand your goal is to minimize the loss from the pullback, but experience has shown that anything trailing with less than 1% will fire WAY too early (price movement is almost never a straight line, and always has some minor (or major) pullbacks.

**Comment by aristosv** [comment]
I'm guessing the next best thing is to configure it like this

```
trailing_stop = True
trailing_stop_positive = 0.01
trailing_stop_positive_offset = 0.02
trailing_only_offset_is_reached = True
```

But then of course `minimal_roi` needs to be set above 2% (or even be disabled?)

```
minimal_roi = {
    "0":  0.02
}
```

One problem I'm seeing here is that with this configuration, each trade will required at least a 2% profit, whereas setting a minimal_roi of 1% would have a better chance of closing trades faster, albeit with less profit.

**Comment by xmatthias** [comment]
it's a tradeoff - either you cap your profits at 2% profit (with ROI) - so you have no chance for a 5% or 10% trade

or you sacrifice the 1% trailing (or more) ... and have the possibility to also have bigger gains.
